controllers/baremetalset_controller.go

```diff
@@ -187,3 +254,322 @@ func (r *BaremetalSetReconciler) provisionServerCreateOrUpdate(instance *ospdire
+	if err != nil {
+		return err
+	}
+
+	// Also remove userdata secret
+	secret := &corev1.Secret{
+		ObjectMeta: v1.ObjectMeta{
+			Name:      fmt.Sprintf(baremetalset.CloudInitSecretName, instance.Name, bmh),
+			Namespace: "openshift-machine-api",
+		},
+	}
+	err = r.Client.Delete(context.Background(), secret, &client.DeleteOptions{})
+	if err != nil {
+		return err
+	}
+	r.Log.Info(fmt.Sprintf("Secret deleted: name %s", secret.Name))
+
+	// Set status (remove this BaremetalHost entry)
+	delete(instance.Status.BaremetalHosts, bmh)
+
+	return nil
+}
+
+func (r *BaremetalSetReconciler) setBaremetalHostStatus(instance *ospdirectorv1beta1.BaremetalSet, bmh *metal3v1alpha1.BareMetalHost, secretName string, mgmtIP string) {
+	// Set status (add this BaremetalHost entry)
+	instance.Status.BaremetalHosts[bmh.GetName()] = ospdirectorv1beta1.BaremetalHostStatus{
+		SecretName: secretName,
+		MgmtIP:     mgmtIP,
+		Online:     bmh.Status.PoweredOn,
+	}
+}
+
+// Deprovision all associated BaremetalHosts for this BaremetalSet via Metal3
+func (r *BaremetalSetReconciler) baremetalHostCleanup(instance *ospdirectorv1beta1.BaremetalSet) error {
+	if instance.Status.BaremetalHosts != nil {
+		for bmhName := range instance.Status.BaremetalHosts {
+			err := r.baremetalHostDeprovision(instance, bmhName)
+
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	return nil
+}
+
+/* deleteOwnerRefLabeledObjects - cleans up namespaced objects outside the default namespace
+   using the owner reference labels added.
+   List of objects which get cleaned:
+   - user-data secret, openshift-machine-api namespace
+*/
+func (r *BaremetalSetReconciler) deleteOwnerRefLabeledObjects(instance *ospdirectorv1beta1.BaremetalSet) error {
+	labelSelectorMap := map[string]string{
+		baremetalset.OwnerUIDLabelSelector:       string(instance.UID),
+		baremetalset.OwnerNameSpaceLabelSelector: instance.Namespace,
+		baremetalset.OwnerNameLabelSelector:      instance.Name,
+	}
+
+	// delete secrets in openshift-machine-api namespace
+	secrets, err := common.GetSecrets(r, "openshift-machine-api", labelSelectorMap)
+	if err != nil {
+		return err
+	}
+	for idx := range secrets.Items {
+		secret := &secrets.Items[idx]
+
+		err = r.Client.Delete(context.Background(), secret, &client.DeleteOptions{})
+		if err != nil {
+			return err
+		}
+		r.Log.Info(fmt.Sprintf("Secret deleted: name %s - %s", secret.Name, secret.UID))
+	}
+
+	return nil
+}
+
+// Get the IP from the management network for a particular BaremetalHost in the BaremetalSet
+func findNewMgmtIP(instance *ospdirectorv1beta1.BaremetalSet, bmhName string) (string, error) {
+	// First check if we already have an IP for this host
+	if bmhStatus, ok := instance.Status.BaremetalHosts[bmhName]; ok {
+		if bmhStatus.MgmtIP != "" {
+			return bmhStatus.MgmtIP, nil
+		}
+	}
+
+	_, network, err := net.ParseCIDR(instance.Spec.MgmtCIDR)
+
+	if err != nil {
+		return "", err
+	}
+
+	for i := uint64(2); i <= cidr.AddressCount(network); i++ {
```

Comment:
I think we want this to be 'i < cidr.AddressCount(network);' as the last IP say 256 in a /24 is reserved

Reply:
Yes, good catch.  The last one is usually the broadcast IP.